cs/src/core/Allocator/AllocatorBase.cs

```diff
@@ -826,29 +826,34 @@ public bool ShiftReadOnlyAddress(long newReadOnlyAddress)
         public void ShiftBeginAddress(long newBeginAddress)
         {
             // First update the begin address
-            var b = Utility.MonotonicUpdate(ref BeginAddress, newBeginAddress, out long oldBeginAddress);
-            b = b && (oldBeginAddress >> LogSegmentSizeBits != newBeginAddress >> LogSegmentSizeBits);
+            if (!Utility.MonotonicUpdate(ref BeginAddress, newBeginAddress, out long oldBeginAddress))
+                return;
 
-            // Then the head address
+            var b = oldBeginAddress >> LogSegmentSizeBits != newBeginAddress >> LogSegmentSizeBits;
+
+            // Shift read-only address
+            epoch.Resume();
+            ShiftReadOnlyAddress(newBeginAddress);
+            epoch.Suspend();
+
+            // Wait for flush to complete
+            while (FlushedUntilAddress < newBeginAddress) ;
+
+            // Then shift head address
             var h = Utility.MonotonicUpdate(ref HeadAddress, newBeginAddress, out long old);
 
-            // Finally the read-only address
-            var r = Utility.MonotonicUpdate(ref ReadOnlyAddress, newBeginAddress, out old);
+            // Ensure no read cache
+            Debug.Assert(!ReadCache);
```

Comment:
Was ReadCache always disallowed when shifting address? Maybe stronger restriction that `Debug.Assert`?

Reply:
Correct, this is not needed; removing.